<a href="https://colab.research.google.com/github/veronicabierhals/CienciaDeDados/blob/main/Arquivos%20Colab/preProcessamentoDados.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Pré-Processamento de Dados**

In [1]:
import pandas as pd

Dados dos jogos de Tenis  
https://raw.githubusercontent.com/JeffSackmann/tennis_atp/master/atp_matches_2021.csv

In [2]:
dados = pd.read_csv('https://raw.githubusercontent.com/JeffSackmann/tennis_atp/master/atp_matches_2021.csv')

Fazer um `read` para entender o tipo de informação que há na tabela.

In [3]:
dados.head(5)

tourney_id    tourney_name surface  draw_size tourney_level  tourney_date  \
0  2021-0096  Tokyo Olympics    Hard         64             A      20210724   
1  2021-0096  Tokyo Olympics    Hard         64             A      20210724   
2  2021-0096  Tokyo Olympics    Hard         64             A      20210724   
3  2021-0096  Tokyo Olympics    Hard         64             A      20210724   
4  2021-0096  Tokyo Olympics    Hard         64             A      20210724   

   match_num  winner_id  winner_seed winner_entry  ... l_1stIn l_1stWon  \
0        237     126207          NaN          NaN  ...     NaN      NaN   
1        238     105526          NaN          NaN  ...     NaN      NaN   
2        239     111576          NaN          NaN  ...     NaN      NaN   
3        240     105357          NaN          NaN  ...     NaN      NaN   
4        241     207830          NaN          NaN  ...     NaN      NaN   

   l_2ndWon l_SvGms  l_bpSaved  l_bpFaced  winner_rank winner_rank_points  \
0       NaN     NaN        NaN        NaN         53.0             1228.0   
1       NaN     NaN        NaN        NaN         48.0             1410.0   
2       NaN     NaN        NaN        NaN        160.0              471.0   
3       NaN     NaN        NaN        NaN         44.0             1476.0   
4       NaN     NaN        NaN        NaN        145.0              525.0   

  loser_rank loser_rank_points  
0       71.0             996.0  
1       95.0             829.0  
2      197.0             354.0  
3       61.0            1106.0  
4      137.0             570.0  

[5 rows x 49 columns]

Verificar a informação dos tipos de cada coluna com `info`

In [4]:
dados.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2733 entries, 0 to 2732
Data columns (total 49 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   tourney_id          2733 non-null   object 
 1   tourney_name        2733 non-null   object 
 2   surface             2733 non-null   object 
 3   draw_size           2733 non-null   int64  
 4   tourney_level       2733 non-null   object 
 5   tourney_date        2733 non-null   int64  
 6   match_num           2733 non-null   int64  
 7   winner_id           2733 non-null   int64  
 8   winner_seed         1165 non-null   float64
 9   winner_entry        417 non-null    object 
 10  winner_name         2733 non-null   object 
 11  winner_hand         2731 non-null   object 
 12  winner_ht           2724 non-null   float64
 13  winner_ioc          2733 non-null   object 
 14  winner_age          2733 non-null   float64
 15  loser_id            2733 non-null   int64  
 16  loser_

Observação de cada um dos atributos com `describe`

In [5]:
dados.describe(include = 'all')

tourney_id     tourney_name surface    draw_size tourney_level  \
count        2733             2733    2733  2733.000000          2733   
unique        114              114       3          NaN             5   
top      2021-580  Australian Open    Hard          NaN             A   
freq          127              127    1608          NaN          1569   
mean          NaN              NaN     NaN    62.609587           NaN   
std           NaN              NaN     NaN    41.521804           NaN   
min           NaN              NaN     NaN     2.000000           NaN   
25%           NaN              NaN     NaN    32.000000           NaN   
50%           NaN              NaN     NaN    64.000000           NaN   
75%           NaN              NaN     NaN   128.000000           NaN   
max           NaN              NaN     NaN   128.000000           NaN   

        tourney_date    match_num      winner_id  winner_seed winner_entry  \
count   2.733000e+03  2733.000000    2733.000000  1165.000000          417   
unique           NaN          NaN            NaN          NaN            6   
top              NaN          NaN            NaN          NaN            Q   
freq             NaN          NaN            NaN          NaN          234   
mean    2.021060e+07   243.248811  127500.755946     7.687554          NaN   
std     2.855522e+02    73.915192   33930.996942     7.133621          NaN   
min     2.021010e+07     1.000000  100644.000000     1.000000          NaN   
25%     2.021032e+07   224.000000  105575.000000     3.000000          NaN   
50%     2.021061e+07   276.000000  106432.000000     6.000000          NaN   
75%     2.021082e+07   289.000000  128034.000000    10.000000          NaN   
max     2.021120e+07   300.000000  210107.000000    32.000000          NaN   

        ...      l_1stIn     l_1stWon     l_2ndWon      l_SvGms    l_bpSaved  \
count   ...  2636.000000  2636.000000  2636.000000  2637.000000  2636.000000   
unique  ...          NaN          NaN          NaN          NaN          NaN   
top     ...          NaN          NaN          NaN          NaN          NaN   
freq    ...          NaN          NaN          NaN          NaN          NaN   
mean    ...    50.118361    33.325493    14.494689    12.420554     4.843323   
std     ...    19.075552    14.205041     6.857036     4.224477     3.253624   
min     ...     2.000000     0.000000     0.000000     0.000000     0.000000   
25%     ...    36.000000    23.000000    10.000000     9.000000     2.000000   
50%     ...    47.000000    31.000000    14.000000    11.000000     4.000000   
75%     ...    61.000000    42.000000    18.000000    15.000000     7.000000   
max     ...   165.000000   104.000000    48.000000    31.000000    27.000000   

          l_bpFaced  winner_rank winner_rank_points   loser_rank  \
count   2636.000000  2729.000000        2729.000000  2707.000000   
unique          NaN          NaN                NaN          NaN   
top             NaN          NaN                NaN          NaN   
freq            NaN          NaN                NaN          NaN   
mean       8.714719    70.593624        2229.913888   103.034355   
std        4.126347   103.028379        2484.822372   149.125316   
min        0.000000     1.000000           1.000000     1.000000   
25%        6.000000    19.000000         858.000000    37.000000   
50%        8.000000    47.000000        1344.000000    68.000000   
75%       11.000000    87.000000        2423.000000   112.000000   
max       38.000000  1758.000000       12113.000000  2146.000000   

       loser_rank_points  
count        2707.000000  
unique               NaN  
top                  NaN  
freq                 NaN  
mean         1390.971555  
std          1460.007475  
min             1.000000  
25%           709.000000  
50%           993.000000  
75%          1595.000000  
max         12113.000000  

[11 rows x 49 columns]

Verificar valores nulos

In [6]:
dados.isnull()

tourney_id  tourney_name  surface  draw_size  tourney_level  \
0          False         False    False      False          False   
1          False         False    False      False          False   
2          False         False    False      False          False   
3          False         False    False      False          False   
4          False         False    False      False          False   
...          ...           ...      ...        ...            ...   
2728       False         False    False      False          False   
2729       False         False    False      False          False   
2730       False         False    False      False          False   
2731       False         False    False      False          False   
2732       False         False    False      False          False   

      tourney_date  match_num  winner_id  winner_seed  winner_entry  ...  \
0            False      False      False         True          True  ...   
1            False      False      False         True          True  ...   
2            False      False      False         True          True  ...   
3            False      False      False         True          True  ...   
4            False      False      False         True          True  ...   
...            ...        ...        ...          ...           ...  ...   
2728         False      False      False         True          True  ...   
2729         False      False      False         True          True  ...   
2730         False      False      False         True          True  ...   
2731         False      False      False         True          True  ...   
2732         False      False      False         True          True  ...   

      l_1stIn  l_1stWon  l_2ndWon  l_SvGms  l_bpSaved  l_bpFaced  winner_rank  \
0        True      True      True     True       True       True        False   
1        True      True      True     True       True       True        False   
2        True      True      True     True       True       True        False   
3        True      True      True     True       True       True        False   
4        True      True      True     True       True       True        False   
...       ...       ...       ...      ...        ...        ...          ...   
2728     True      True      True     True       True       True        False   
2729     True      True      True     True       True       True        False   
2730     True      True      True     True       True       True        False   
2731     True      True      True     True       True       True        False   
2732     True      True      True     True       True       True        False   

      winner_rank_points  loser_rank  loser_rank_points  
0                  False       False              False  
1                  False       False              False  
2                  False       False              False  
3                  False       False              False  
4                  False       False              False  
...                  ...         ...                ...  
2728               False       False              False  
2729               False       False              False  
2730               False       False              False  
2731               False       False              False  
2732               False       False              False  

[2733 rows x 49 columns]

Verificar valores nulos nas colunas

In [7]:
dados.isnull().sum()

tourney_id               0
tourney_name             0
surface                  0
draw_size                0
tourney_level            0
tourney_date             0
match_num                0
winner_id                0
winner_seed           1568
winner_entry          2316
winner_name              0
winner_hand              2
winner_ht                9
winner_ioc               0
winner_age               0
loser_id                 0
loser_seed            2040
loser_entry           2098
loser_name               0
loser_hand               0
loser_ht                35
loser_ioc                0
loser_age                0
score                    0
best_of                  0
round                    0
minutes                 83
w_ace                   97
w_df                    97
w_svpt                  97
w_1stIn                 97
w_1stWon                97
w_2ndWon                97
w_SvGms                 96
w_bpSaved               97
w_bpFaced               97
l_ace                   97
l

Coluna "winner_entry" tem muitos valores nulos e indica como o vencedor entrou no torneio.

In [11]:
dados['winner_entry'].describe()

count     417
unique      6
top         Q
freq      234
Name: winner_entry, dtype: object

In [12]:
dados['winner_entry'].unique()

array([nan, 'Q', 'WC', 'LL', 'SE', 'Alt', 'PR'], dtype=object)

Apagar todas as linhas que o atributo "loser_rank" é nulo, pois são poucas ocasiões e isso pode implicar erros nos resultados.

In [14]:
dados.dropna(subset = ['loser_rank'], inplace = True)
print(dados)

     tourney_id    tourney_name surface  draw_size tourney_level  \
0     2021-0096  Tokyo Olympics    Hard         64             A   
1     2021-0096  Tokyo Olympics    Hard         64             A   
2     2021-0096  Tokyo Olympics    Hard         64             A   
3     2021-0096  Tokyo Olympics    Hard         64             A   
4     2021-0096  Tokyo Olympics    Hard         64             A   
...         ...             ...     ...        ...           ...   
2728  2021-9210       Laver Cup    Hard          8             A   
2729  2021-9210       Laver Cup    Hard          8             A   
2730  2021-9210       Laver Cup    Hard          8             A   
2731  2021-9210       Laver Cup    Hard          8             A   
2732  2021-9210       Laver Cup    Hard          8             A   

      tourney_date  match_num  winner_id  winner_seed winner_entry  ...  \
0         20210724        237     126207          NaN          NaN  ...   
1         20210724        238    

Resultado: de 2733 linhas passou para 2707 linhas.

Apagar colunas que todos os valores são nulos.
obs: neste caso não há colunas com todos valores nulos.

In [15]:
dados.dropna(axis = 1, how = 'all')

tourney_id    tourney_name surface  draw_size tourney_level  \
0     2021-0096  Tokyo Olympics    Hard         64             A   
1     2021-0096  Tokyo Olympics    Hard         64             A   
2     2021-0096  Tokyo Olympics    Hard         64             A   
3     2021-0096  Tokyo Olympics    Hard         64             A   
4     2021-0096  Tokyo Olympics    Hard         64             A   
...         ...             ...     ...        ...           ...   
2728  2021-9210       Laver Cup    Hard          8             A   
2729  2021-9210       Laver Cup    Hard          8             A   
2730  2021-9210       Laver Cup    Hard          8             A   
2731  2021-9210       Laver Cup    Hard          8             A   
2732  2021-9210       Laver Cup    Hard          8             A   

      tourney_date  match_num  winner_id  winner_seed winner_entry  ...  \
0         20210724        237     126207          NaN          NaN  ...   
1         20210724        238     105526          NaN          NaN  ...   
2         20210724        239     111576          NaN          NaN  ...   
3         20210724        240     105357          NaN          NaN  ...   
4         20210724        241     207830          NaN          NaN  ...   
...            ...        ...        ...          ...          ...  ...   
2728      20210924        102     126610          NaN          NaN  ...   
2729      20210924        103     106421          NaN          NaN  ...   
2730      20210924        104     126094          NaN          NaN  ...   
2731      20210924        105     100644          NaN          NaN  ...   
2732      20210924        106     134770          NaN          NaN  ...   

     l_1stIn l_1stWon  l_2ndWon l_SvGms  l_bpSaved  l_bpFaced  winner_rank  \
0        NaN      NaN       NaN     NaN        NaN        NaN         53.0   
1        NaN      NaN       NaN     NaN        NaN        NaN         48.0   
2        NaN      NaN       NaN     NaN        NaN        NaN        160.0   
3        NaN      NaN       NaN     NaN        NaN        NaN         44.0   
4        NaN      NaN       NaN     NaN        NaN        NaN        145.0   
...      ...      ...       ...     ...        ...        ...          ...   
2728     NaN      NaN       NaN     NaN        NaN        NaN          7.0   
2729     NaN      NaN       NaN     NaN        NaN        NaN          2.0   
2730     NaN      NaN       NaN     NaN        NaN        NaN          5.0   
2731     NaN      NaN       NaN     NaN        NaN        NaN          4.0   
2732     NaN      NaN       NaN     NaN        NaN        NaN         10.0   

     winner_rank_points loser_rank loser_rank_points  
0                1228.0       71.0             996.0  
1                1410.0       95.0             829.0  
2                 471.0      197.0             354.0  
3                1476.0       61.0            1106.0  
4                 525.0      137.0             570.0  
...                 ...        ...               ...  
2728             5173.0       11.0            3368.0  
2729            10575.0       12.0            3310.0  
2730             6130.0       15.0            2800.0  
2731             7760.0       22.0            2171.0  
2732             3440.0       19.0            2341.0  

[2707 rows x 49 columns]

Alterar valor dos dados nulos em uma coluna.

In [16]:
dados['loser_entry'].unique()

array([nan, 'PR', 'Q', 'LL', 'WC', 'SE', 'Alt'], dtype=object)

Substitui nulos por valor "x".
Utilizado especificamente para valores nulos.

In [17]:
dados['loser_entry'].fillna(value = 'x', inplace = True)

In [18]:
dados['loser_entry'].unique()

array(['x', 'PR', 'Q', 'LL', 'WC', 'SE', 'Alt'], dtype=object)

Alterar mais de um atributos, com condições diferentes.

In [19]:
dados['w_ace'].unique()

array([nan,  5.,  1.,  2.,  3.,  6.,  4., 11.,  8., 15.,  9., 20., 12.,
       24., 14., 17., 13., 19., 10.,  7., 16., 18., 22.,  0., 23., 21.,
       32., 26., 25., 27., 28., 29., 41., 34., 30., 44., 33., 49., 31.,
       36.])

In [20]:
dados['loser_ht'].unique()

array([180., 183., 188., 185., 193., 191., 173., 170., 178., 196., 198.,
       175., 181., 211., 190., 208., 203.,  nan, 172., 206., 194., 201.,
       189.])

`'loser_ht': dados['loser_ht'].mean()` faz uma média dos valores que não são nulos e preenche com essa média.

In [21]:
dados.fillna(value = {'w_ace': 0, 'loser_ht': dados['loser_ht'].mean()}, inplace = True)

In [22]:
dados['w_ace'].unique()

array([ 0.,  5.,  1.,  2.,  3.,  6.,  4., 11.,  8., 15.,  9., 20., 12.,
       24., 14., 17., 13., 19., 10.,  7., 16., 18., 22., 23., 21., 32.,
       26., 25., 27., 28., 29., 41., 34., 30., 44., 33., 49., 31., 36.])

In [23]:
dados['loser_ht'].unique()

array([180.        , 183.        , 188.        , 185.        ,
       193.        , 191.        , 173.        , 170.        ,
       178.        , 196.        , 198.        , 175.        ,
       181.        , 211.        , 190.        , 208.        ,
       203.        , 186.63866171, 172.        , 206.        ,
       194.        , 201.        , 189.        ])

Usar o método `ffill` para projetar um dado para o registro seguinte.  
Se um dados é nulo, pega o dado anterior e coloca no dado nulo.

In [31]:
dados['minutes'][1975:1990]

1978     67.0
1979    129.0
1980     91.0
1981    140.0
1982     83.0
1983     74.0
1984     90.0
1985     71.0
1986    124.0
1987    140.0
1988    169.0
1989     59.0
1990    130.0
1991    116.0
1992    102.0
Name: minutes, dtype: float64

In [33]:
dados['minutes'].fillna(method = 'ffill', inplace = True)

In [34]:
dados['minutes'][1975:1990]

1978     67.0
1979    129.0
1980     91.0
1981    140.0
1982     83.0
1983     74.0
1984     90.0
1985     71.0
1986    124.0
1987    140.0
1988    169.0
1989     59.0
1990    130.0
1991    116.0
1992    102.0
Name: minutes, dtype: float64

Identificar duplicados

In [35]:
dados.duplicated()

0       False
1       False
2       False
3       False
4       False
        ...  
2728    False
2729    False
2730    False
2731    False
2732    False
Length: 2707, dtype: bool

In [36]:
dados.duplicated().sum()

0

Eliminar duplicados de uma coluna (cuidado ao executar).

In [37]:
dados.drop_duplicates(subset = ['tourney_name'], inplace = True)

In [38]:
dados['tourney_name'].head(10)

0      Tokyo Olympics
64             Munich
91       Queen's Club
122            Gstaad
149           Newport
176            Bastad
203         Kitzbuhel
230         Stuttgart
257            Geneva
284            Vienna
Name: tourney_name, dtype: object

In [39]:
dados.describe()

draw_size  tourney_date   match_num      winner_id  winner_seed  \
count  110.000000  1.100000e+02  110.000000     110.000000    11.000000   
mean    28.272727  2.021077e+07  147.463636  132444.963636     2.636364   
std     31.388045  3.299224e+02  127.788578   35819.636900     2.540580   
min      2.000000  2.021010e+07    1.000000  100644.000000     1.000000   
25%      4.000000  2.021050e+07    1.000000  105606.250000     1.000000   
50%     32.000000  2.021092e+07  239.000000  124940.500000     2.000000   
75%     32.000000  2.021111e+07  271.000000  136022.500000     2.000000   
max    128.000000  2.021120e+07  286.000000  209867.000000     9.000000   

        winner_ht  winner_age       loser_id  loser_seed    loser_ht  ...  \
count  108.000000  110.000000     110.000000    2.000000  110.000000  ...   
mean   185.814815   26.620909  127675.763636    7.500000  186.035600  ...   
std      5.913944    4.900966   36480.909687    0.707107    6.694571  ...   
min    170.000000   17.800000  103852.000000    7.000000  170.000000  ...   
25%    183.000000   23.550000  105526.000000    7.250000  183.000000  ...   
50%    185.000000   25.100000  106165.500000    7.500000  185.000000  ...   
75%    188.000000   29.525000  130971.750000    7.750000  189.500000  ...   
max    198.000000   40.100000  208502.000000    8.000000  211.000000  ...   

          l_1stIn    l_1stWon    l_2ndWon     l_SvGms   l_bpSaved   l_bpFaced  \
count  106.000000  106.000000  106.000000  106.000000  106.000000  106.000000   
mean    46.113208   29.518868   12.481132   10.933962    4.462264    8.141509   
std     19.663028   13.125933    5.926906    3.002440    3.604691    4.681145   
min     16.000000    5.000000    2.000000    4.000000    0.000000    1.000000   
25%     33.250000   20.250000    8.000000    9.000000    2.000000    5.000000   
50%     42.500000   27.000000   12.500000   10.000000    4.000000    7.500000   
75%     56.000000   37.000000   16.000000   13.000000    6.000000   11.000000   
max    165.000000   90.000000   32.000000   18.000000   27.000000   38.000000   

       winner_rank  winner_rank_points   loser_rank  loser_rank_points  
count   109.000000          109.000000   110.000000         110.000000  
mean    130.963303         1659.550459   206.081818         774.890909  
std     167.072003         2475.617400   276.922608         722.899353  
min       1.000000            7.000000     5.000000           1.000000  
25%      42.000000          477.000000    67.000000         238.000000  
50%      95.000000          811.000000   101.500000         780.000000  
75%     155.000000         1492.000000   251.750000         999.000000  
max    1165.000000        12113.000000  1921.000000        5150.000000  

[8 rows x 35 columns]

# **Como fazer deslocamento**

Dados das ações da Apple
https://finance.yahoo.com/quote/AAPL/history/

In [42]:
dados2 = pd.read_csv("AAPL.csv")

In [43]:
dados2

Date        Open        High         Low       Close   Adj Close  \
0    2023-01-25  140.889999  142.429993  138.809998  141.860001  141.071487   
1    2023-01-26  143.169998  144.250000  141.899994  143.960007  143.159805   
2    2023-01-27  143.160004  147.229996  143.080002  145.929993  145.118851   
3    2023-01-30  144.960007  145.550003  142.850006  143.000000  142.205139   
4    2023-01-31  142.699997  144.339996  142.279999  144.289993  143.487961   
..          ...         ...         ...         ...         ...         ...   
246  2024-01-18  186.089996  189.139999  185.830002  188.630005  188.630005   
247  2024-01-19  189.330002  191.949997  188.820007  191.559998  191.559998   
248  2024-01-22  192.300003  195.330002  192.259995  193.889999  193.889999   
249  2024-01-23  195.020004  195.750000  193.830002  195.179993  195.179993   
250  2024-01-24  195.419998  196.380005  194.339996  194.500000  194.500000   

       Volume  
0    65799300  
1    54105100  
2    70555800  
3    64015300  
4    65874500  
..        ...  
246  78005800  
247  68741000  
248  60133900  
249  42355600  
250  53578100  

[251 rows x 7 columns]

In [45]:
dados2['close_ontem'] = dados2['Close'].shift(1)

In [46]:
dados2

Date        Open        High         Low       Close   Adj Close  \
0    2023-01-25  140.889999  142.429993  138.809998  141.860001  141.071487   
1    2023-01-26  143.169998  144.250000  141.899994  143.960007  143.159805   
2    2023-01-27  143.160004  147.229996  143.080002  145.929993  145.118851   
3    2023-01-30  144.960007  145.550003  142.850006  143.000000  142.205139   
4    2023-01-31  142.699997  144.339996  142.279999  144.289993  143.487961   
..          ...         ...         ...         ...         ...         ...   
246  2024-01-18  186.089996  189.139999  185.830002  188.630005  188.630005   
247  2024-01-19  189.330002  191.949997  188.820007  191.559998  191.559998   
248  2024-01-22  192.300003  195.330002  192.259995  193.889999  193.889999   
249  2024-01-23  195.020004  195.750000  193.830002  195.179993  195.179993   
250  2024-01-24  195.419998  196.380005  194.339996  194.500000  194.500000   

       Volume  close_ontem  
0    65799300          NaN  
1    54105100   141.860001  
2    70555800   143.960007  
3    64015300   145.929993  
4    65874500   143.000000  
..        ...          ...  
246  78005800   182.679993  
247  68741000   188.630005  
248  60133900   191.559998  
249  42355600   193.889999  
250  53578100   195.179993  

[251 rows x 8 columns]

Nova coluna com a porcentagem de alteração de um dia para o outro.

In [47]:
dados2['alteracao %'] = ((dados2['Close']/dados2['close_ontem'])- 1)*100

In [48]:
dados2.head(20)

Date        Open        High         Low       Close   Adj Close  \
0   2023-01-25  140.889999  142.429993  138.809998  141.860001  141.071487   
1   2023-01-26  143.169998  144.250000  141.899994  143.960007  143.159805   
2   2023-01-27  143.160004  147.229996  143.080002  145.929993  145.118851   
3   2023-01-30  144.960007  145.550003  142.850006  143.000000  142.205139   
4   2023-01-31  142.699997  144.339996  142.279999  144.289993  143.487961   
5   2023-02-01  143.970001  146.610001  141.320007  145.429993  144.621628   
6   2023-02-02  148.899994  151.179993  148.169998  150.820007  149.981689   
7   2023-02-03  148.029999  157.380005  147.830002  154.500000  153.641220   
8   2023-02-06  152.570007  153.100006  150.779999  151.729996  150.886612   
9   2023-02-07  150.639999  155.229996  150.639999  154.649994  153.790375   
10  2023-02-08  153.880005  154.580002  151.169998  151.919998  151.075562   
11  2023-02-09  153.779999  154.330002  150.419998  150.869995  150.031403   
12  2023-02-10  149.460007  151.339996  149.220001  151.009995  150.399887   
13  2023-02-13  150.949997  154.259995  150.919998  153.850006  153.228439   
14  2023-02-14  152.119995  153.770004  150.860001  153.199997  152.581055   
15  2023-02-15  153.110001  155.500000  152.880005  155.330002  154.702438   
16  2023-02-16  153.509995  156.330002  153.350006  153.710007  153.089005   
17  2023-02-17  152.350006  153.000000  150.850006  152.550003  151.933685   
18  2023-02-21  150.199997  151.300003  148.410004  148.479996  147.880127   
19  2023-02-22  148.869995  149.949997  147.160004  148.910004  148.308380   

       Volume  close_ontem  alteracao %  
0    65799300          NaN          NaN  
1    54105100   141.860001     1.480337  
2    70555800   143.960007     1.368426  
3    64015300   145.929993    -2.007807  
4    65874500   143.000000     0.902093  
5    77663600   144.289993     0.790076  
6   118339000   145.429993     3.706260  
7   154357300   150.820007     2.439990  
8    69858300   154.500000    -1.792883  
9    83322600   151.729996     1.924470  
10   64120100   154.649994    -1.765274  
11   56007100   151.919998    -0.691155  
12   57450700   150.869995     0.092795  
13   62199000   151.009995     1.880678  
14   61707600   153.850006    -0.422495  
15   65573800   153.199997     1.390343  
16   68167900   155.330002    -1.042938  
17   59144100   153.710007    -0.754670  
18   58867200   152.550003    -2.667982  
19   51011300   148.479996     0.289607

In [49]:
dados2.dropna(subset = ['alteracao %'], inplace = True)

In [50]:
dados2

Date        Open        High         Low       Close   Adj Close  \
1    2023-01-26  143.169998  144.250000  141.899994  143.960007  143.159805   
2    2023-01-27  143.160004  147.229996  143.080002  145.929993  145.118851   
3    2023-01-30  144.960007  145.550003  142.850006  143.000000  142.205139   
4    2023-01-31  142.699997  144.339996  142.279999  144.289993  143.487961   
5    2023-02-01  143.970001  146.610001  141.320007  145.429993  144.621628   
..          ...         ...         ...         ...         ...         ...   
246  2024-01-18  186.089996  189.139999  185.830002  188.630005  188.630005   
247  2024-01-19  189.330002  191.949997  188.820007  191.559998  191.559998   
248  2024-01-22  192.300003  195.330002  192.259995  193.889999  193.889999   
249  2024-01-23  195.020004  195.750000  193.830002  195.179993  195.179993   
250  2024-01-24  195.419998  196.380005  194.339996  194.500000  194.500000   

       Volume  close_ontem  alteracao %  
1    54105100   141.860001     1.480337  
2    70555800   143.960007     1.368426  
3    64015300   145.929993    -2.007807  
4    65874500   143.000000     0.902093  
5    77663600   144.289993     0.790076  
..        ...          ...          ...  
246  78005800   182.679993     3.257068  
247  68741000   188.630005     1.553302  
248  60133900   191.559998     1.216330  
249  42355600   193.889999     0.665323  
250  53578100   195.179993    -0.348393  

[250 rows x 9 columns]

In [51]:
dados2.describe(include = 'all')

Date        Open        High         Low       Close  \
count          250  250.000000  250.000000  250.000000  250.000000   
unique         250         NaN         NaN         NaN         NaN   
top     2023-01-26         NaN         NaN         NaN         NaN   
freq             1         NaN         NaN         NaN         NaN   
mean           NaN  175.662801  177.228680  174.409920  175.961200   
std            NaN   14.470799   14.257612   14.431214   14.288183   
min            NaN  142.699997  144.250000  141.320007  143.000000   
25%            NaN  166.092499  167.942497  165.544994  166.709995   
50%            NaN  177.230003  179.250000  176.230004  177.565002   
75%            NaN  187.909996  189.260002  186.929996  188.075001   
max            NaN  198.020004  199.619995  197.000000  198.110001   

         Adj Close        Volume  close_ontem  alteracao %  
count   250.000000  2.500000e+02   250.000000   250.000000  
unique         NaN           NaN          NaN          NaN  
top            NaN           NaN          NaN          NaN  
freq           NaN           NaN          NaN          NaN  
mean    175.580855  5.817773e+07   175.750640     0.134321  
std      14.459460  1.728123e+07    14.401306     1.268248  
min     142.205139  2.404830e+07   141.860001    -4.802005  
25%     166.150143  4.747335e+07   166.514999    -0.655255  
50%     177.331382  5.419000e+07   177.525002     0.121955  
75%     187.913201  6.478352e+07   188.047497     0.843269  
max     198.110001  1.543573e+08   198.110001     4.692692

# Tranformar os dados em categóricos(qualitativos) em dados numéricos.

In [52]:
dados = pd.read_csv('https://raw.githubusercontent.com/JeffSackmann/tennis_atp/master/atp_matches_2021.csv')

País de origem do jogador.

In [53]:
dados['winner_ioc']

0       USA
1       GER
2       IND
3       AUS
4       CZE
       ... 
2728    ITA
2729    RUS
2730    RUS
2731    GER
2732    NOR
Name: winner_ioc, Length: 2733, dtype: object

Verificar valores únicos.

In [54]:
dados['winner_ioc'].unique()

array(['USA', 'GER', 'IND', 'AUS', 'CZE', 'KAZ', 'SRB', 'BLR', 'COL',
       'CRO', 'FRA', 'GBR', 'GEO', 'ESP', 'ITA', 'RUS', 'ARG', 'POL',
       'JPN', 'GRE', 'URU', 'LTU', 'NOR', 'SVK', 'CAN', 'BEL', 'NED',
       'AUT', 'SWE', 'CHI', 'RSA', 'TPE', 'FIN', 'SUI', 'DEN', 'HUN',
       'BUL', 'SLO', 'BIH', 'POR', 'ECU', 'PER', 'BRA', 'BOL', 'UKR',
       'MDA', 'TUN', 'KOR', 'TUR', 'ISR', 'ROU', 'BAR', 'INA', 'DOM',
       'MEX', 'ESA', 'PAR', 'THA'], dtype=object)

Ver documentação do scikit-learn sobre preprocessing  
https://scikit-learn.org/stable/auto_examples/index.html#preprocessing

In [55]:
from sklearn.preprocessing import LabelEncoder

In [58]:
paises = dados['winner_ioc']
paises

0       USA
1       GER
2       IND
3       AUS
4       CZE
       ... 
2728    ITA
2729    RUS
2730    RUS
2731    GER
2732    NOR
Name: winner_ioc, Length: 2733, dtype: object

Gerar países por números.

In [59]:
lencoder = LabelEncoder()
paises_numeros = lencoder.fit_transform(paises)
paises_numeros

array([57, 24, 28, ..., 45, 24, 38])

In [60]:
paises.head(10)

0    USA
1    GER
2    IND
3    AUS
4    CZE
5    KAZ
6    GER
7    SRB
8    USA
9    BLR
Name: winner_ioc, dtype: object

Gerar uma nova coluna com o número dos países.

In [61]:
dados['paises_numeros'] = paises_numeros
dados

tourney_id    tourney_name surface  draw_size tourney_level  \
0     2021-0096  Tokyo Olympics    Hard         64             A   
1     2021-0096  Tokyo Olympics    Hard         64             A   
2     2021-0096  Tokyo Olympics    Hard         64             A   
3     2021-0096  Tokyo Olympics    Hard         64             A   
4     2021-0096  Tokyo Olympics    Hard         64             A   
...         ...             ...     ...        ...           ...   
2728  2021-9210       Laver Cup    Hard          8             A   
2729  2021-9210       Laver Cup    Hard          8             A   
2730  2021-9210       Laver Cup    Hard          8             A   
2731  2021-9210       Laver Cup    Hard          8             A   
2732  2021-9210       Laver Cup    Hard          8             A   

      tourney_date  match_num  winner_id  winner_seed winner_entry  ...  \
0         20210724        237     126207          NaN          NaN  ...   
1         20210724        238     105526          NaN          NaN  ...   
2         20210724        239     111576          NaN          NaN  ...   
3         20210724        240     105357          NaN          NaN  ...   
4         20210724        241     207830          NaN          NaN  ...   
...            ...        ...        ...          ...          ...  ...   
2728      20210924        102     126610          NaN          NaN  ...   
2729      20210924        103     106421          NaN          NaN  ...   
2730      20210924        104     126094          NaN          NaN  ...   
2731      20210924        105     100644          NaN          NaN  ...   
2732      20210924        106     134770          NaN          NaN  ...   

     l_1stWon l_2ndWon  l_SvGms l_bpSaved  l_bpFaced  winner_rank  \
0         NaN      NaN      NaN       NaN        NaN         53.0   
1         NaN      NaN      NaN       NaN        NaN         48.0   
2         NaN      NaN      NaN       NaN        NaN        160.0   
3         NaN      NaN      NaN       NaN        NaN         44.0   
4         NaN      NaN      NaN       NaN        NaN        145.0   
...       ...      ...      ...       ...        ...          ...   
2728      NaN      NaN      NaN       NaN        NaN          7.0   
2729      NaN      NaN      NaN       NaN        NaN          2.0   
2730      NaN      NaN      NaN       NaN        NaN          5.0   
2731      NaN      NaN      NaN       NaN        NaN          4.0   
2732      NaN      NaN      NaN       NaN        NaN         10.0   

      winner_rank_points loser_rank loser_rank_points paises_numeros  
0                 1228.0       71.0             996.0             57  
1                 1410.0       95.0             829.0             24  
2                  471.0      197.0             354.0             28  
3                 1476.0       61.0            1106.0              1  
4                  525.0      137.0             570.0             14  
...                  ...        ...               ...            ...  
2728              5173.0       11.0            3368.0             30  
2729             10575.0       12.0            3310.0             45  
2730              6130.0       15.0            2800.0             45  
2731              7760.0       22.0            2171.0             24  
2732              3440.0       19.0            2341.0             38  

[2733 rows x 50 columns]